sdk/python/kfp/compiler/compiler.py

```diff
@@ -182,7 +184,86 @@ def _get_condition_params_for_ops_helper(group, current_conditions_params):
     _get_condition_params_for_ops_helper(root_group, [])
     return conditions
 
-  def _get_inputs_outputs(self, pipeline, root_group, op_groups, opsgroup_groups, condition_params):
+  def _get_next_group_or_op(cls, to_visit: List, already_visited: Set):
+    """Get next group or op to visit."""
+    if len(to_visit) == 0:
+      return None
+    next = to_visit.pop(0)
+    while next in already_visited:
+      next = to_visit.pop(0)
+    already_visited.add(next)
+    return next
+
+  def _get_for_loop_ops(self, new_root) -> Dict[Text, dsl.ParallelFor]:
+    to_visit = self._get_subgroups(new_root)
+    op_name_to_op = {}
+    already_visited = set()
+
+    while len(to_visit):
+      next_op = self._get_next_group_or_op(to_visit, already_visited)
+      if next_op is None:
+        break
+      to_visit.extend(self._get_subgroups(next_op))
+      if isinstance(next_op, dsl.ParallelFor):
+        op_name_to_op[next_op.name] = next_op
+
+    return op_name_to_op
+
+  def _fill_loop_args(self, new_root):
```

Comment:
Could you move all the related logic in another module file for better readability? This class has already been too big.

Reply:
removing method